### PR TITLE
Specifications

### LINKED ISSUE #154: Move ISpecification<> from example-shipping into main EventFlow lib
Ditto

### PR DESCRIPTION
**Pull request by rasmus**
@JC008 What do you think? I wanted developers to be able to provide information regarding why a specification isn't satisfied.

Fixes #154

### CODE REVIEW

**Comment by JC008** [comment]
@rasmus Looks good and very DDD + the multi-reason handles that age old issue of giving **complete** feedback in one step.

How would you handle a `set` of `ISpecification`?  E.g. where:

- all must satisfy
- at least one must satisfy
- at most `n` must satisfy
- at least `n` must satisfy

File: Source/EventFlow/Extensions/SpeficicationExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+using System;
+using System.Linq;
+using EventFlow.Exceptions;
+using EventFlow.Provided.Specifications;
+using EventFlow.Specifications;
+
+namespace EventFlow.Extensions
+{
+    public static class SpeficicationExtensions
```
Comment (JC008) [inline Source/EventFlow/Extensions/SpeficicationExtensions.cs]: Typo - SpeficicationExtensions

File: Source/EventFlow/Provided/Specifications/AggregateIsNewSpecification.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Collections.Generic;
+using EventFlow.Aggregates;
+using EventFlow.Specifications;
+
+namespace EventFlow.Provided.Specifications
+{
+    public class AggregateIsNewSpecification : Specification<IAggregateRoot>
```
Comment (JC008) [inline Source/EventFlow/Provided/Specifications/AggregateIsNewSpecification.cs]: Awesome!!

**Comment by rasmus** [comment]
@JC008 What do you think of the solution for the all, and at least `n` specifications?

File: Source/EventFlow.Tests/UnitTests/Specifications/SpecificationTests.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+            // Act
+            var isSatisfiedBy = isTrue.IsSatisfiedBy(false);
+
+            // Act
+            isSatisfiedBy.Should().BeFalse();
+        }
+
+        [Test]
+        public void NotSpeficication_ReturnsTrue_ForNotSatisfied()
```
Comment (JC008) [inline Source/EventFlow.Tests/UnitTests/Specifications/SpecificationTests.cs]: Typo - NotSpeficication_

**Comment by JC008** [comment]
@rasmus Very nice - this allows for complex specifications (business rules/cases) to be applied and composed with Linq e.g. `if customer has purchased product X in time frame Y as well as subscribed to the newsletter and gave more than two instances of feedback` he can get a 5% discount!

Also - look at the [repo.FindBy](https://lostechies.com/jimmybogard/2007/10/25/specifications-versus-validators/) example which shows how specifications could be used in `IQueryHandler`.

**Pull request merged**